tower/src/retry/policy.rs

```diff
@@ -56,6 +56,21 @@ pub trait Policy<Req, Res, E>: Sized {
     /// If the request *should* be retried, return `Some` future of a new
     /// policy that would apply for the next request attempt.
     ///
+    /// ## Mutating Requests
+    ///
+    /// The policy MAY chose to mutate the `req`: if the request is mutated, the mutated request
+    /// will be sent to the inner service in the next retry. This can be helpful for use cases like
+    /// tracking the retry count in a header.
+    ///
+    /// ## Mutating Results
+    ///
+    /// The policy MAY chose to mutate the result. This can enable the retry policy to convert a failure
+    /// into a success and vice versa. For example, if the policy is used to poll while waiting for a state
+    /// change, you can switch the result to failure such that running out of retries returns a specific failure.
```

Comment:
nit: in general, we prefer to avoid second-person ('you') language in docs:
```suggestion
    /// change, the policy can switch the result to failure such that running out of retries returns a specific failure.
```